File: csharp/Repository/Database/UserDatabaseRepository.cs
using Domain.Models;
using log4net;
using Repository.Interfaces;
using Repository.Utils;
using System.Data.Common;

namespace Repository.Database
{
    public class UserDatabaseRepository : IUserRepository
    {
        private DatabaseUtils databaseUtils;


        private ILog log = LogManager.GetLogger(typeof(UserDatabaseRepository));

        private static readonly String INSERT_USER_SQL_STRING = "insert into users(username, password) values(@username_param, @password_param);";
        private static readonly String FIND_BY_USERNAME_SQL_STRING = "select user_id, username, password from users where username=@username_param";

        public UserDatabaseRepository(IDictionary<String, String> databaseConnectionProperties)
        {
            log.Info("Creating user database repository...");
            databaseUtils = new DatabaseUtils(databaseConnectionProperties);
            log.Info("Created user dabase repository!");
        }

        public User Add(User model)
        {
            log.Info("Adding user " + model.ToString());

            try
            {
                using (var connection = databaseUtils.GetConnection())
                using (var insertCommand = connection.CreateCommand())
                {
                    insertCommand.CommandText = INSERT_USER_SQL_STRING;

                    var usernameParam = insertCommand.CreateParameter();
                    usernameParam.ParameterName = "@username_param";
                    usernameParam.Value = model.UserName;

                    var passwordParam = insertCommand.CreateParameter();
                    passwordParam.ParameterName = "@password_param";
                    passwordParam.Value = model.Password;

                    insertCommand.Parameters.Add(usernameParam);
                    insertCommand.Parameters.Add(passwordParam);

                    insertCommand.ExecuteNonQuery();

                    User? addedUser = GetByUsername(model.UserName);
                    if (addedUser == null)
                        throw new DatabaseException("Couldn't add user " + addedUser);
                    return addedUser;
                }
            }
            catch(DbException ex)
            {
                log.Error(ex.ToString());
                throw new DatabaseException("Error occured while adding user " + model);
            }

        }
        public User? GetByUsername(string username)
        {
            try
            {
                using (var connection = databaseUtils.GetConnection())
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = FIND_BY_USERNAME_SQL_STRING;

                    var usernameParam = command.CreateParameter();
                    usernameParam.ParameterName = "@username_param";
                    usernameParam.Value = username;

                    command.Parameters.Add(usernameParam);

                    using (var reader = command.ExecuteReader())
                    {
                        if (reader.Read())
                        {
                            long userId = reader.GetInt64(0);
                            String foundUsername = reader.GetString(1);
                            String password = reader.GetString(2);
                            User user = new User(username, password);
                            user.Id = userId;
                            return user;
                        }
                        else
                            return null;
                    }
                }
            } 
            catch (DbException ex)
            {
                log.Error(ex.Message);
                throw new DatabaseException("Error occured when searching for user " + username);
            }
        }

        public ICollection<User> GetAll()
        {
            throw new NotImplementedException();
        }

        public User GetById(long id)
        {
            throw new NotImplementedException();
        }

    }
}

File: java/repository/src/main/java/database/UserDatabaseRepository.java
package database;

import exceptions.DatabaseException;
import exceptions.NotImplementedMethodException;
import interfaces.UserRepository;
import models.User;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import utils.JdbcUtils;

import java.sql.SQLException;
import java.sql.Statement;
import java.util.Optional;
import java.util.Properties;
import java.util.function.Function;


public class UserDatabaseRepository implements UserRepository {
    private JdbcUtils utils;
    private Logger log = LogManager.getLogger();

    private static final String INSERT_USER_SQL_STRING = "insert into users(username, password) values(?, ?);";
    private static final String FIND_BY_USERNAME_SQL_STRING = "select user_id, username, password from users where username=?";

    public UserDatabaseRepository(Properties databaseConnectionProperties) {
        log.traceEntry("Creating user database repository...");
        utils = new JdbcUtils(databaseConnectionProperties);
        log.traceExit();
    }

    @Override
    public User save(User model) {
        Function<User, User> saveUserFunction = user -> {
            log.traceEntry("Adding new user " + model);

            var existingUserOptional = findUserByUserName(model.getUserName());
            if (existingUserOptional.isPresent()) {
                log.info("User {} already exists", model.getUserName());
                log.traceExit();
                return existingUserOptional.get();
            }

            try (var connection = utils.getConnection();
                 var insertPreparedStatement = connection.prepareStatement(INSERT_USER_SQL_STRING, Statement.RETURN_GENERATED_KEYS)) {

                insertPreparedStatement.setString(1, model.getUserName());
                insertPreparedStatement.setString(2, model.getPassword());

                insertPreparedStatement.executeUpdate();

                try (var generatedKeys = insertPreparedStatement.getGeneratedKeys()) {
                    if (generatedKeys.next()) {
                        User addedUser = new User(model);
                        addedUser.setId(generatedKeys.getLong(1));
                        log.info("Added user {}", addedUser.getUserName());
                        return addedUser;
                    }
                }
            } catch (SQLException exception) {
                log.error(exception);
            }
            log.traceExit();
            return null;
        };

        var savedUser = saveUserFunction.apply(model);

        if (savedUser == null)
            throw new DatabaseException("DatabaseException occurred while saving user {} " +  model.getUserName());

        return savedUser;
    }

    @Override
    public Optional<User> findUserByUserName(String username) {
        Function<String, Optional<User>> findUserFunction = usrname -> {
            log.traceEntry("Searching for user with username {}", username);
            try (var connection = utils.getConnection();
                 var findPreparedStatement = connection.prepareStatement(FIND_BY_USERNAME_SQL_STRING)) {

                findPreparedStatement.setString(1, username);

                try (var resultSet = findPreparedStatement.executeQuery()) {
                    if (resultSet.next()) {
                        var userId = resultSet.getLong(1);
                        var existingUsername = resultSet.getString(2);
                        var password = resultSet.getString(3);
                        User user = new User(existingUsername, password);
                        user.setId(userId);
                        log.info("Found user");
                        log.traceExit();
                        return Optional.of(user);
                    } else {
                        log.info("User not found");
                        log.traceExit();
                        return Optional.empty();
                    }
                }
            } catch (SQLException exception) {
                log.error(exception);
            }
            log.traceExit();
            return null;
        };

        var foundUserOptional = findUserFunction.apply(username);

        if (foundUserOptional == null)
            throw new DatabaseException("DatabaseException occurred while searching for user with username " + username);

        return foundUserOptional;
    }

    @Override
    public User findOne(Long modelID) {
        throw new NotImplementedMethodException();
    }

    @Override
    public Iterable<User> findAll() {
        throw new NotImplementedMethodException();
    }
}
